### PR TITLE
Store failure messages for outgoing payments

### PR DESCRIPTION
**Pull request by rorp**
Currently there's no way to get payment failures using `getsentinfo` RPC. This PR addresses that.

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/db/PaymentsDb.scala [hunk 1]
```diff
@@ -26,7 +26,7 @@ trait PaymentsDb {
   def addOutgoingPayment(outgoingPayment: OutgoingPayment)
 
   // updates the status of the payment, if the newStatus is SUCCEEDED you must supply a preimage
-  def updateOutgoingPayment(id: UUID, newStatus: OutgoingPaymentStatus.Value, preimage: Option[ByteVector32] = None)
+  def updateOutgoingPayment(id: UUID, newStatus: OutgoingPaymentStatus.Value, preimage: Option[ByteVector32] = None, failures: Traversable[String] = Seq())
```
Comment (araspitzu) [inline eclair-core/src/main/scala/fr/acinq/eclair/db/PaymentsDb.scala]: Why is this a `Traversable[String]`  instead of `Seq[String]`? For consistency it should be a `Seq` as it's defined in `OutgoingPayment`

File: eclair-core/src/main/scala/fr/acinq/eclair/db/sqlite/SqlitePaymentsDb.scala [hunk 2]
```diff
@@ -18,27 +18,33 @@ package fr.acinq.eclair.db.sqlite
...
+
 import fr.acinq.bitcoin.ByteVector32
 import fr.acinq.eclair.db.sqlite.SqliteUtils._
 import fr.acinq.eclair.db.{IncomingPayment, OutgoingPayment, OutgoingPaymentStatus, PaymentsDb}
 import fr.acinq.eclair.payment.PaymentRequest
 import grizzled.slf4j.Logging
+
 import scala.collection.immutable.Queue
 import OutgoingPaymentStatus._
+
```
Comment (araspitzu) [inline eclair-core/src/main/scala/fr/acinq/eclair/db/sqlite/SqlitePaymentsDb.scala]: nit: please remove blank lines

File: eclair-core/src/main/scala/fr/acinq/eclair/db/sqlite/SqlitePaymentsDb.scala [hunk 3]
```diff
@@ -65,30 +73,34 @@ class SqlitePaymentsDb(sqlite: Connection) extends PaymentsDb with Logging {
...
   override def getOutgoingPayment(id: UUID): Option[OutgoingPayment] = {
-    using(sqlite.prepareStatement("SELECT id, payment_hash, preimage, amount_msat, created_at, completed_at, status FROM sent_payments WHERE id = ?")) { statement =>
+    val res = using(sqlite.prepareStatement("SELECT id, payment_hash, preimage, amount_msat, created_at, completed_at, status FROM sent_payments WHERE id = ?")) { statement =>
       statement.setString(1, id.toString)
       val rs = statement.executeQuery()
       if (rs.next()) {
+        val id = UUID.fromString(rs.getString("id"))
         Some(OutgoingPayment(
-          UUID.fromString(rs.getString("id")),
+          id,
```
Comment (araspitzu) [inline eclair-core/src/main/scala/fr/acinq/eclair/db/sqlite/SqlitePaymentsDb.scala]: nit: unnecessary change
```suggestion
         UUID.fromString(rs.getString("id")),
```

File: eclair-core/src/main/scala/fr/acinq/eclair/db/sqlite/SqlitePaymentsDb.scala [hunk 4]
```diff
@@ -225,4 +241,26 @@ class SqlitePaymentsDb(sqlite: Connection) extends PaymentsDb with Logging {
...
+    }
+  }
+
+  private def selectFailures(id: UUID): Seq[String] = {
+    val res = new mutable.ListBuffer[String]
+    using(sqlite.prepareStatement("SELECT failure FROM sent_payments_failures WHERE id = ?")) { statement =>
+      statement.setString(1, id.toString)
+      val rs = statement.executeQuery()
+      while (rs.next()) {
+        res += rs.getString(1)
```
Comment (araspitzu) [inline eclair-core/src/main/scala/fr/acinq/eclair/db/sqlite/SqlitePaymentsDb.scala]: nit: for consistency getters use string based parameters
```suggestion
        res += rs.getString("failure")
```

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentLifecycle.scala [hunk 5]
```diff
@@ -49,13 +49,13 @@ class PaymentLifecycle(nodeParams: NodeParams, id: UUID, router: ActorRef, regis
   when(WAITING_FOR_REQUEST) {
     case Event(c: SendPaymentToRoute, WaitingForRequest) =>
       val send = SendPayment(c.amountMsat, c.paymentHash, c.hops.last, finalCltvExpiry = c.finalCltvExpiry, maxAttempts = 1)
-      paymentsDb.addOutgoingPayment(OutgoingPayment(id, c.paymentHash, None, c.amountMsat, Platform.currentTime, None, OutgoingPaymentStatus.PENDING))
+      paymentsDb.addOutgoingPayment(OutgoingPayment(id, c.paymentHash, None, c.amountMsat, Platform.currentTime, None, OutgoingPaymentStatus.PENDING, Seq()))
```
Comment (araspitzu) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentLifecycle.scala]: nit: please use the literal empty sequence
```suggestion
      paymentsDb.addOutgoingPayment(OutgoingPayment(id, c.paymentHash, None, c.amountMsat, Platform.currentTime, None, OutgoingPaymentStatus.PENDING, Seq.empty))
```

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentLifecycle.scala [hunk 6]
```diff
@@ -49,13 +49,13 @@ class PaymentLifecycle(nodeParams: NodeParams, id: UUID, router: ActorRef, regis
...
       router ! FinalizeRoute(c.hops)
       goto(WAITING_FOR_ROUTE) using WaitingForRoute(sender, send, failures = Nil)
 
     case Event(c: SendPayment, WaitingForRequest) =>
       router ! RouteRequest(nodeParams.nodeId, c.targetNodeId, c.amountMsat, c.assistedRoutes, routeParams = c.routeParams)
-      paymentsDb.addOutgoingPayment(OutgoingPayment(id, c.paymentHash, None, c.amountMsat, Platform.currentTime, None, OutgoingPaymentStatus.PENDING))
+      paymentsDb.addOutgoingPayment(OutgoingPayment(id, c.paymentHash, None, c.amountMsat, Platform.currentTime, None, OutgoingPaymentStatus.PENDING, Seq()))
```
Comment (araspitzu) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentLifecycle.scala]: Same as above.

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/Relayer.scala [hunk 7]
```diff
@@ -137,7 +137,7 @@ class Relayer(nodeParams: NodeParams, register: ActorRef, paymentHandler: ActorR
         case Local(id, None) =>
           // we sent the payment, but we probably restarted and the reference to the original sender was lost,
           // we publish the failure on the event stream and update the status in paymentDb
-          nodeParams.db.payments.updateOutgoingPayment(id, OutgoingPaymentStatus.FAILED)
+          nodeParams.db.payments.updateOutgoingPayment(id, OutgoingPaymentStatus.FAILED, failures = Nil)
```
Comment (araspitzu) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/Relayer.scala]: This can become `Seq.empty` once the parameter becomes of type `Seq`

File: eclair-core/src/test/scala/fr/acinq/eclair/db/SqlitePaymentsDbSpec.scala [hunk 8]
```diff
@@ -138,10 +139,11 @@ class SqlitePaymentsDbSpec extends FunSuite {
     val s3 = s2.copy(id = UUID.randomUUID(), amountMsat = 88776655)
     db.addOutgoingPayment(s3)
 
-    db.updateOutgoingPayment(s3.id, FAILED)
+    db.updateOutgoingPayment(s3.id, FAILED, failures = Seq("one", "two", "three"))
```
Comment (araspitzu) [inline eclair-core/src/test/scala/fr/acinq/eclair/db/SqlitePaymentsDbSpec.scala]: Can you also try to insert a record with the exactly same failure? It should work because the PRIMARY KEY for `sent_payments_failures` is not defined and it should use the `ROWID` field of sqlite, however if the primary key was defined as the compound of the two columns an attacker could return the same error twice causing a constraint violation when inserting the record.

**Comment by rorp** [comment]
We can create a separate DB for remotely generated strings and/or use something like `org.apache.commons.text.StringEscapeUtils` to sanitize them.

**Comment by t-bast** [comment]
This change has been included in #1130 :tada:
